Add org support for role None

buccarel opened the pull request:
Git card:
https://github.com/grafana/identity-access-team/issues/302

In the UI, the new user looks as follows:
<img width="949" alt="image" src="https://github.com/grafana/terraform-provider-grafana/assets/11213785/59746be1-4abb-4ece-91a6-c3853dcac3ae">
Not sure if anything should be done here to display "None" in the "Role" box.

In the DB, the org_user looks as follows:
<img width="591" alt="image" src="https://github.com/grafana/terraform-provider-grafana/assets/11213785/564f9030-a647-48ee-8e12-b14c0bffaa37">

The target version >= 10.2.0 is not yet configured in the CI, so to see that the new unit test is running, I did this locally, against the latest Grafana `main` branch:
```
mattia.buccarella at Mattias-Grafana-Macbook in ~/workspace/terraform-provider-grafana (buccarel/support-role-none)
$ GRAFANA_VERSION=10.2.0 TF_ACC_OSS=true GRAFANA_URL=http://localhost:3000 GRAFANA_AUTH=admin:admin TESTARGS="-run TestAccOrganization_roleNoneUser" make testacc | grep roleNone
TF_ACC=1 go test ./... -v -run TestAccOrganization_roleNoneUser -timeout 120m
=== RUN   TestAccOrganization_roleNoneUser
--- PASS: TestAccOrganization_roleNoneUser (1.82s)

mattia.buccarella at Mattias-Grafana-Macbook in ~/workspace/terraform-provider-grafana (buccarel/support-role-none)
$
```

Commit:
Add org support for role None

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/terraform-provider-grafana?pullRequest=1068) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/terraform-provider-grafana?pullRequest=1068) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: julienduchesne<br/>:x: buccarel<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/terraform-provider-grafana?pullRequest=1068) it.</sub>

Commit:
Regenerating docs

eleijonmarck commented on docs/resources/organization.md:
```diff
@@ -67,6 +67,9 @@ already exist in Grafana.
 - `editors` (Set of String) A list of email addresses corresponding to users who should be given editor
 access to the organization. Note: users specified here must already exist in
 Grafana unless 'create_users' is set to true.
+- `nones` (Set of String) A list of email addresses corresponding to users who has no basic role assigned.
```
```suggestion
- `nones` (Set of String) A list of email addresses corresponding to users who should be given the none access to the organization.
```

eleijonmarck commented on internal/resources/grafana/resource_organization.go:
```diff
@@ -125,6 +125,18 @@ Grafana unless 'create_users' is set to true.
+					Type: schema.TypeString,
+				},
+				Description: `
+A list of email addresses corresponding to users who has no basic role assigned.
```
```suggestion
A list of email addresses corresponding to users who who should be given none access to the organization.
```

buccarel committed:
Update docs/resources/organization.md

Co-authored-by: Eric Leijonmarck <eric.leijonmarck@gmail.com>

buccarel committed:
Update internal/resources/grafana/resource_organization.go

Co-authored-by: Eric Leijonmarck <eric.leijonmarck@gmail.com>

Commit:
Regenerating docs

julienduchesne commented on internal/resources/grafana/resource_organization_test.go:
```diff
@@ -110,6 +113,9 @@ func TestAccOrganization_users(t *testing.T) {
 						"grafana_organization.test", "viewers.#",
 					),
+					resource.TestCheckNoResourceAttr(
+						"grafana_organization.test", "nones.#",
```
I think this will need a test where there is one in the list

buccarel replied:
Sounds good!
Will do, thanks!

julienduchesne commented on docs/resources/organization.md:
```diff
@@ -67,6 +67,9 @@ already exist in Grafana.
 - `editors` (Set of String) A list of email addresses corresponding to users who should be given editor
 access to the organization. Note: users specified here must already exist in
 Grafana unless 'create_users' is set to true.
+- `nones` (Set of String) A list of email addresses corresponding to users who who should be given none access to the organization.
```
I would call it: `users_without_access` maybe. The others are self descriptive, this one, not so much. A 1:1 match with the API is not required

Commit:
Testing assignment and removal of role none

Commit:
making ad-hoc test

eleijonmarck commented:
@buccarel @julienduchesne this will be targeting the release of Grafana 10.2. So we have to merge this after the 10.2 release of Grafana.

@julienduchesne could you confirm this to me?

eleijonmarck commented:
@buccarel on another note, do you know why the "None" role do not show up as a set value to your image.
![image](https://github.com/grafana/terraform-provider-grafana/assets/7727602/622f14d6-12af-4b21-a5f7-94cbe8db8e06)

buccarel commented:
> @buccarel on another note, do you know why the "None" role do not show up as a set value to your image. ![image](https://user-images.githubusercontent.com/7727602/273594562-622f14d6-12af-4b21-a5f7-94cbe8db8e06.png)

In the description of the PR I guessed that the UI doesn't have the validation of the new role, but in reality I didn't check why this happens. After I finishing addressing all of the comments, I'll try to investigate it.

buccarel commented:
> @buccarel @julienduchesne this will be targeting the release of Grafana 10.2. So we have to merge this after the 10.2 release of Grafana.
> 
> @julienduchesne could you confirm this to me?

With my understanding of the picture, yes: I already thought this PR should be merged once that version is released.

Commit:
Rename list from nones to users_without_access

Commit:
Regen docs and map roles to role lists

Commit:
More generic functions to obtain role name and role list name

IevaVasiljeva commented on docs/resources/organization.md:
```diff
@@ -67,6 +67,9 @@ already exist in Grafana.
 - `editors` (Set of String) A list of email addresses corresponding to users who should be given editor
 access to the organization. Note: users specified here must already exist in
 Grafana unless 'create_users' is set to true.
+- `users_without_access` (Set of String) A list of email addresses corresponding to users who who should be given none access to the organization.
```
nit:
```suggestion
- `users_without_access` (Set of String) A list of email addresses corresponding to users who should be given none access to the organization.
```

IevaVasiljeva commented on internal/resources/grafana/resource_organization.go:
```diff
@@ -125,6 +125,18 @@ Grafana unless 'create_users' is set to true.
+					Type: schema.TypeString,
+				},
+				Description: `
+A list of email addresses corresponding to users who who should be given none access to the organization.
```
```suggestion
A list of email addresses corresponding to users who should be given none access to the organization.
```

IevaVasiljeva reviewed (approved):
Changes look good to me! But I agree with you that we need to wait until Grafana 10.2 is released before merging this :)

buccarel committed:
Update docs/resources/organization.md

Co-authored-by: Ieva <ieva.vasiljeva@grafana.com>

buccarel committed:
Update internal/resources/grafana/resource_organization.go

Co-authored-by: Ieva <ieva.vasiljeva@grafana.com>

vtorosyan commented:
@eleijonmarck pinging you on this one, can you please follow up when capacity allows to ensure we get this change merged as well?

julienduchesne committed:
Fix semver check + add 10.2 notice
